### ISSUE TITLE
Sync does not create folders / directories in root of destination drive on Local FS

### ISSUE THREAD

**Issue by varian93** [ncw/rclone#2451]
#### What is the problem you are having with rclone?
When syncing Rclone is unable to create folders in the root of a destination drive on the local file system. Consequently it can not copy any files into these folders as they do not exist.

I do not have any sort of cloud storage setup to verify if this is global behavior or limited to the local file system. 

#### What is your rclone version (eg output from `rclone -V`)
I discovered the issue while using: v1.42
I then switched to the latest beta: v1.42-083-gd95667b0 
The issue persisted.

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Windows 10 64 bit

####  Which cloud storage system are you using? (eg Google Drive)
No cloud = Local File System

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
I did many tests to isolate what the exact issue was that was causing the problem of files not being copied. The last test I did that clarified the issue included the command:

rclone --log-file=rclone_%date%.log --log-level=DEBUG --stats-file-name-length 0 --retries 1 --delete-after --exclude="$RECYCLE.BIN/" --exclude="System Volume Information/" --exclude="pagefile.sys" sync Z:\ Y:\

In this test I created the folder "Z:\Data" which contained 1 file "info.txt". The file was not empty. It contained some text. The folder "Data" did not yet exist on drive "Y". All other folders located in the root of drive "Z" already existed in the root of drive "Y" before beginning the sync. I tested for case sensitive issues. The folder being named "data" did not differ the result.

It seems that Rclone has no issue with creating folders in any other location than the root of the destination drive. It will properly create and remove folders in any other location.

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

Due to the size of the drive I was attempting to sync the resulting log file was over 30MB in size. Thus I've only posted what I thought were the relevant parts of the log file. If I missed something that is needed please let me know.

Log File Contents:

2018/08/08 10:13:28 DEBUG : rclone: Version "v1.42-083-gd95667b0" starting with parameters ["rclone" "--log-file=rclone_2018-08-08.log" "--log-level=DEBUG" "--stats-file-name-length" "0" "--retries" "1" "--delete-after" "--exclude=$RECYCLE.BIN/" "--exclude=System Volume Information/" "--exclude=pagefile.sys" "sync" "Z:\\" "Y:\\"]
2018/08/08 10:13:28 DEBUG : $RECYCLE.BIN: Excluded from sync (and deletion)
2018/08/08 10:13:28 DEBUG : $RECYCLE.BIN: Excluded from sync (and deletion)
2018/08/08 10:13:28 DEBUG : System Volume Information: Excluded from sync (and deletion)
2018/08/08 10:13:28 DEBUG : System Volume Information: Excluded from sync (and deletion)
2018/08/08 10:13:28 ERROR : Data/info.txt: Failed to copy: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
2018/08/08 10:13:33 INFO  : Local file system at \\?\Y:\: Waiting for transfers to finish
2018/08/08 10:13:33 DEBUG : Local file system at \\?\Y:\: failed to copy 1 directories
2018/08/08 10:13:33 DEBUG : Local file system at \\?\Y:\: copied 270 directories
2018/08/08 10:13:33 ERROR : Local file system at \\?\Y:\: not deleting files as there were IO errors
2018/08/08 10:13:33 ERROR : Local file system at \\?\Y:\: not deleting directories as there were IO errors
2018/08/08 10:13:33 ERROR : Attempt 1/1 failed with 1 errors and: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
2018/08/08 10:13:33 Failed to sync: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.

**Comment by ncw** [comment]
I think this is probably the same issue as #1754 which is due to be fixed by go1.11 (released at the end of Aug 18).

Can you try the workaround in that thread and see if it fixes it for you?

I can compile a newer version of rclone with go1.11 easily enough.

**Comment by ncw** [comment]
I've been compiling [the latest beta](https://beta.rclone.org) with go1.11rc1 so this issue should be fixed there.

I'm planning to release 1.43 after go1.11 is released so this issue should be fixed in a release then.

**Comment by varian93** [comment]
I tried using:

rclone-beta-latest-windows-amd64.zip | 6.7 MiB | 08/21/2018, 4:55:45 PM

from the link in your post: https://beta.rclone.org/

It failed. I then looked at the version (rclone -V) and it output the following:

rclone v1.42-126-g3751ceeb
- os/arch: windows/amd64
- go version: go1.10

Did I download the wrong file or did you make a mistake in compiling?

**Comment by ncw** [comment]
Hmm, I discovered the problem eventually! Appveyour hadn't updated their go version.

They have now - can you have a go with this?

https://beta.rclone.org/branch/v1.42-148-g77dca018-test/

Thanks

**Comment by varian93** [comment]
I tested the version that you linked to and it worked as we expected, however I discovered another, what seems to be, extremely closely related bug while testing it.

The version tested:

**rclone v1.42-148-g77dca018-test
-os/arch: windows/amd64
-go version: go1.11**

In my test I created a couple of folders in the root of the source that did not exist on the destination. I also created a test file in the root of the source that did not exist on the destination. It successfully handled the directories that did not exist on the destination, it however failed on the file.

I thought that this new behavior that I noticed may have just recently emerged, so I pulled up the old version that you had linked to in #1754 that I've been using that has been working:

**rclone v1.40-DEV
-os/arch: windows/amd64
-go version: devel +fdf33730e1 Sun Apr 1 06:55:54 2018 +0000**

It behaved the same way. It handles everything as expected _except_ for files in the root of the source.

It becomes more complicated. I experimented to see if it would work if the file already existed on the destination but had just been modified on the source. It failed.

Simply put, any files in the root of the source will fail when being synced to the root of the destination. 

The only exception to this that I've been able to prove is if the **only** thing that changed about the file is it's modification time. In this case rclone simply states that it updated it's modification time on the destination. 

I never previously ran into this problem because the system I have in place contains **no** files in the root. Just to make my tests comprehensive I included files in the root while testing.

Again, I don't know if this problem exists for any other system than the local file system. I have no cloud systems implemented to test with.

It's my assumption that these issues are only recently being discovered because I'm syncing "root to root", Z:\ to Y:\

I imagine most people are syncing root to folders or folders to folders.


**Pertinent Details:**

rclone v1.42-148-g77dca018-test
-os/arch: windows/amd64
-go version: go1.11

**The following error results if the file already exists on the destination or not:**

2018/08/30 14:32:22 ERROR : New Text Document.txt: Failed to copy: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
2018/08/30 14:32:23 INFO  : Local file system at \\?\Y:\: Waiting for checks to finish
2018/08/30 14:32:23 INFO  : Local file system at \\?\Y:\: Waiting for transfers to finish
2018/08/30 14:32:23 ERROR : Local file system at \\?\Y:\: not deleting files as there were IO errors
2018/08/30 14:32:23 ERROR : Local file system at \\?\Y:\: not deleting directories as there were IO errors
2018/08/30 14:32:23 ERROR : Attempt 1/1 failed with 1 errors and: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
2018/08/30 14:32:23 Failed to sync: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.


**If the only thing that has changed is the file's modification time the error log files reads as thus:**

2018/08/30 14:35:13 INFO  : New Text Document.txt: Updated modification time in destination
2018/08/30 14:35:14 INFO  : Local file system at \\?\Y:\: Waiting for checks to finish
2018/08/30 14:35:14 INFO  : Local file system at \\?\Y:\: Waiting for transfers to finish
2018/08/30 14:35:14 INFO  : Waiting for deletions to finish
2018/08/30 14:35:14 INFO  : 
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:            119434 / 119434, 100%
Transferred:            0 / 0, -
Elapsed time:        1.5s


I tried to be as clear as possible. I hope I've been sufficiently successful.

**Comment by ncw** [comment]
@varian93 thanks for testing.

I'm glad this issue is working :-)  I managed to replicate the additional issue

```
mkdir C:\src
subst P: C:\src
# Create a file called z.txt in C:\src
rclone --retries 1 -vv copy P:\ C:\
```

gives

```
2018/08/31 23:50:22 DEBUG : rclone: Version "v1.42-DEV" starting with parameters ["rclone" "--retries" "1" "-vv" "copy" "P:\\" "C:\\"]
2018/08/31 23:50:22 DEBUG : Using config file from "C:\\Users\\Dev\\.rclone.conf"
2018/08/31 23:50:22 NOTICE: Documents and Settings: Can't follow symlink without -L/--copy-links
2018/08/31 23:50:22 INFO  : Local file system at \\?\C:\: Waiting for checks tofinish
2018/08/31 23:50:22 DEBUG : z.txt: Sizes differ (src 5 vs dst 9)
2018/08/31 23:50:22 INFO  : Local file system at \\?\C:\: Waiting for transfers to finish
2018/08/31 23:50:22 ERROR : z.txt: Failed to copy: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
2018/08/31 23:50:22 ERROR : Attempt 1/1 failed with 1 errors and: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
2018/08/31 23:50:22 Failed to copy: mkdir \\?: The filename, directory name, or volume label syntax is incorrect.
```

I think this is a bug in the go runtime, but not 100% sure.  I managed to put a workaround in though which I'm happy with, can you try this - it works for me now.

https://beta.rclone.org/v1.42-156-gd49ba652/ (uploaded in 15-30 mins)

**Comment by varian93** [comment]
No problem, I don't mind testing.

I tried the version that you linked to in your previous post & it works.

I could not get it to fail.

Thank you for implementing the workaround. I'm glad we got it figured out, now to get it to work without having to implement a workaround...

**Comment by ncw** [comment]
Thank you for testing - glad it works for you!

I'm happy with the work-around all it is is making sure I never pass `\\?\C:` to mkdir instead of `\\?\C:\` which I probably should have been doing anyway.

So I'm going to close this now as fixed :-)

The fix went into 1.43 :-)
